core/src/main/scala/org/apache/spark/shuffle/ShuffleWriter.scala

```diff
@@ -17,18 +17,466 @@
+   * same shuffle will map shuffle partition ranges to individual target locations in a consistent
+   * manner to make sure each target location receives shuffle blocks belonging to the same set
+   * of partition ranges. 0-length blocks and blocks that are large enough will be skipped.
+   *
+   * @param numPartitions Number of shuffle partitions in the shuffle file
+   * @param partitionId map index of the current mapper
+   * @param shuffleId ShuffleId of current shuffle
+   * @param dataFile Shuffle data file
+   * @param partitionLengths array of sizes of blocks in the shuffle data file
+   * @param mergerLocs Target locations to push blocks to
+   * @param transportConf TransportConf used to create FileSegmentManagedBuffer
+   * @param maxBlockSizeToPush Max size of individual blocks that will be pushed. Blocks larger
+   *                           than this threshold will be skipped.
+   * @param maxBlockBatchSize Max size of a batch of shuffle blocks to be grouped into a single
+   *                          request
+   * @return List of the PushRequest, randomly shuffled.
+   */
+  private def prepareBlockPushRequests(
+      numPartitions: Int,
+      partitionId: Int,
+      shuffleId: Int,
+      dataFile: File,
+      partitionLengths: Array[Long],
+      mergerLocs: Seq[BlockManagerId],
+      transportConf: TransportConf,
+      maxBlockSizeToPush: Long,
+      maxBlockBatchSize: Long): Seq[PushRequest] = {
+    var offset = 0L
+    var currentReqSize = 0L
+    var currentReqOffset = 0L
+    var currentMergerId = 0
+    val numMergers = mergerLocs.length
+    val requests = new ArrayBuffer[PushRequest]
+    var blocks = new ArrayBuffer[(BlockId, Long)]
+    for (reduceId <- 0 until numPartitions) {
+      val blockSize = partitionLengths(reduceId)
+      logDebug(
+        s"Block ${ShufflePushBlockId(shuffleId, partitionId, reduceId)} is of size $blockSize")
+      // Skip 0-length blocks and blocks that are large enough
+      if (blockSize > 0) {
+        val mergerId = math.min(math.floor(reduceId * 1.0 / numPartitions * numMergers),
+          numMergers - 1).asInstanceOf[Int]
+        // Start a new PushRequest if the current request goes beyond the max batch size,
+        // or the number of blocks in the current request goes beyond the limit per destination,
+        // or the next block push location is for a different shuffle service, or the next block
+        // exceeds the max block size to push limit. This guarantees that each PushReqeust
+        // represents continuous blocks in the shuffle file to be pushed to the same shuffle
+        // service, and does not go beyond existing limitations.
+        if (currentReqSize + blockSize <= maxBlockBatchSize
+          && blocks.size < maxBlocksInFlightPerAddress
+          && mergerId == currentMergerId && blockSize <= maxBlockSizeToPush) {
+          // Add current block to current batch
+          currentReqSize += blockSize
+        } else {
```

Comment:
This is to make sure we are grouping contiguous blocks into each batch. When we read the block data from disk, we actually combine all the blocks in one batch into a single I/O to make the I/O more efficient. If we skip, it would then break this guarantee.

Reply:
I see. But for the case like [10, 1000, 10, 1000, 100] (assuming the `maxBlockSizeToPush=800`), the I/O efficiency might also be low?

Reply:
@Victsm

Reply:
I left another comment about this max block size configuration, which might be relevant to this one. as well. https://github.com/apache/spark/pull/30312#discussion_r530575752

We could have the following scenarios:

1. All blocks are large enough. In this case, we simply do not push the blocks.
2. All blocks are small enough. In this case, we would always get perfect batch size.
3. Majority of the blocks are small enough, with a small percentage being large. In this case, we could have a few batches not reaching the ideal batch size. However, majority of the batches will still have ideal size, thus improving I/O efficiency.
4. Majority of the blocks are large enough, with a small percentage being small. In this case, we might be just pushing  these few small blocks. Since they are few, it also doesn't hurt overall I/O efficiency.
5. The block size is a mix, with small and large blocks scattered in the shuffle file. This is where a static threshold does not perform well enough, and the proposal I mentioned in the other comment might help. Though in reality, this scenario is much less likely to happen compared with the first 4 scenarios.